### DISCUSSION TITLE
Feature Request: Automatically pick (latest) patch version for Go

### DISCUSSION THREAD

**Discussion by phil9909** [containerbase/base#3085] [Ideas]
When I run `install-tool golang 1.22`  I get the following:

```bash
[07:12:03.727] INFO (5): Installing tool golang@1.22...
Invalid version: 1.22
[07:12:03.809] ERROR (5): Command failed with exit code 1: /usr/local/containerbase/bin/install-tool.sh golang 1.22
[07:12:03.810] FATAL (5): Install tool golang failed in 83ms.
```

What I would hope to get instead: Installing the latest available patch version of 1.22 (as of today that would be 1.22.6).

This feature could be widened to support installing the latest patch version of the latest minor version ( `install-tool golang 1` would currently install 1.22.6 but soon 1.23.0 instead). Leaving the version out would give you the latest available (as already supported for node). It could further be expanded to support arbitrary semver constraints like `>=1 <2`, but I'm not sure how useful that would be.

Obviously there is no reason to limit this to the `golang` tool. `install-tool node` for example will give you the latest available (LTS I guess) or you can use `install-tool node x.y.z` to get an exact version. But there is no easy way to get the latest patch for `v21`.

See also #3016. I'm willing to contribute this again in TypeScript.
Also, let me know if you want help with the conversion of the golang tool to TypeScript.

# Use Case

We use this in [renovate `postUpgradeTasks`](https://docs.renovatebot.com/configuration-options/#postupgradetasks) to re-run the [protobuf compiler](https://github.com/protocolbuffers/protobuf) when it is updated. And use `grep '^go ' go.mod | cut -d' ' -f2` to extract the version (`major.minor` format).

Everywhere else (e.g. in our Dockerfiles) we use the latest available patch of golang and would like to keep that consistent.

**Comment by viceice** [comment]
the tool will soon be able to detect the version for the tools from eg go.mod for go and use it to fetch a proper version. see open issues..

- so first step is to convert to typescript.
- then we can add a resolver to get a go version as you like.
